### PR TITLE
[8.0] port server_monitoring

### PR DESCRIPTION
**Pull request by gurneyalex**

### CODE REVIEW

File: server_monitoring/__openerp__.py
Comment (RoelAdriaans) [inline server_monitoring/__openerp__.py]: Description can be removed because you have a README.rst file

File: server_monitoring/__openerp__.py
Comment (RoelAdriaans) [inline server_monitoring/__openerp__.py]: No longer needed to have a depends on base in v8 modules

Comment (gurneyalex) [inline server_monitoring/__openerp__.py]: fixed

File: server_monitoring/README.rst
Comment (tafaRU) [inline server_monitoring/README.rst]: Please remove three above lines, see https://github.com/OCA/maintainer-tools/pull/140 as reference.

File: server_monitoring/__openerp__.py
Comment (tafaRU) [inline server_monitoring/__openerp__.py]: Please update version according to https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

Comment (lmignon) [inline server_monitoring/__openerp__.py]: Is-it the right scheme for version umber? IMO it should be 8.0.1.0.0 https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

**Comment by StefanRijnhart** [comment]
Tested succesfully. Great module! Some process info entries have no model set, with common methods such as fields_view_get or read. Is this expected?

**Comment by StefanRijnhart** [comment]
Now, if I log out of the database with this module installed and log on to another database on the same instance I get 

```
server-tools/server_monitoring/models/server_monitor_process.py", line 109, in _call_kw
    monitor_obj.log_measure(request.cr, request.uid, model, method,
AttributeError: 'NoneType' object has no attribute 'log_measure'
```

and I can not use this database.

**Comment by gurneyalex** [comment]
@StefanRijnhart probably something I missed when adding the json rpc support. I'll check. 

re: multiple db, I admit I never test it. I'll check it.

**Comment by gurneyalex** [comment]
@guewen you mentionned you had a solution for the monkey patching to only happen when the module is actually installed. Can you tell me more about this?

**Comment by sebalix** [comment]
@gurneyalex Maybe the `post_load` key in the manifest: https://github.com/odoo/odoo/blob/8.0/openerp/modules/module.py#L322
OCA's template module mentions it regarding monkey patching: https://github.com/OCA/maintainer-tools/blob/master/template/module/hooks.py#L49

**Comment by guewen** [comment]
I didn't know `post_load`, it could be a clean solution!

The one I had in head was the `_register_hook` (https://github.com/OCA/server-tools/blob/b2251a7ed3a12533bcf304e2a456fc3be000024c/base_import_match/models/base_import.py#L70). But it needs to be done on a model (it would be an arbitrary model in your case) so less readable/understandable than the `post_load` solution (if `post_load` works as I expect :).

**Comment by dreispt** [comment]
This PR is about to make it's first birthday! 😸

**Comment by gurneyalex** [comment]
@dreispt it's much older than this... it had started as a MP on launchpad on v7...

File: server_monitoring/__openerp__.py
Comment (hbrunn) [inline server_monitoring/__openerp__.py]: no `base`?

**Review by hbrunn — APPROVED** [review]
only small things

**Comment by hbrunn** [comment]
one word of caution about `post_load`: This is run once server wide, so your monkey patches can end up being installed because the module is installed in database A, but run for B and then it will break. Here, I think we need to be quite sure that the module is actually installed, so you should follow the pattern of using `_register_hook` while checking if you already did the monkey patching (`_register_hook` can be called more often than once). Auditlog does that extensively: https://github.com/OCA/server-tools/blob/8.0/auditlog/models/rule.py#L104
I'm curious if you ever measured the performance penalty of the module? And if this is non-negligible, would it be useful to be able to turn that off for a running server?
When this is merged, I'll want to marry it to dead_mans_switch so that we can pull all the data into our own odoo's dashboard if we want.

File: server_monitoring/__openerp__.py
Comment (RoelAdriaans) [reply server_monitoring/__openerp__.py]: Base is no longer needed as a dependency in the newest odoo.

File: server_monitoring/models/server_monitor_process.py
Comment (hbrunn) [inline server_monitoring/models/server_monitor_process.py]: the extra underscore messes up stuff I guess

**Comment by gurneyalex** [comment]
@hbrunn I've updated the module following your suggestion (and using a naive lock). There is some additional protection on the monkey patched version to do nothing  when the model is not in the registry. 

Regarding the perf. hit, I admit I have no measures, but there must be some since each RPC call triggers a write. This can be mitigated by:
- adding a configuration option to globally not log RPC access (or maybe later we could want to fine tune this on a per-model basis)
- I'm also thinking of a cheap optimization in which the logs would be kept in memory for some time, and only flushed after a number of RPC calls. This could mean loosing some logs when the instance is shut down but I can probably live with it.

**Comment by hbrunn** [comment]
@gurneyalex thanks for your points, you just need to loose one underscore I think.

About the performance stuff, sounds good. Parameters are read through a registry wide cache anyways, so this won't impact performance itself when checking whether to log or lot. But let's not block on this and take it as a future extension. I'm especially looking forward to having a convenient way of comparing index usage across different DBs and customer profiles.

**Comment by guewen** [comment]
> adding a configuration option to globally not log RPC access (or maybe later we could want to fine tune this on a per-model basis)

Fine-tuning might be done by overrides of the `log_measures` methods. A global switch might be useful, but not strictly needed to merge this pull request.

> I'm also thinking of a cheap optimization in which the logs would be kept in memory for some time, and only flushed after a number of RPC calls. This could mean loosing some logs when the instance is shut down but I can probably live with it.

I also losing some logs is not an issue, we usually want to measure the tendencies. Not needed too for a merge.

**Comment by gurneyalex** [comment]
@hbrunn @guewen I just added a configuration setting for enabling RPC calls monitoring.

**Comment by hbrunn** [comment]
I still approve ;-) But I removed the needs fixing part as the error on travis seems quite unrelated

**Comment by moylop260** [comment]
FYI the [lint build](https://travis-ci.org/OCA/server-tools/jobs/161299646#L441-L442) show us:

``` bash
server_monitoring/models/server_monitor_process.py:91:19: F821 undefined name 'self'
server_monitoring/models/field_types.py:29: [C8104(class-camelcase), bigint] Use `CamelCase` "Bigint" in class name "bigint"...
```

**Comment by hbrunn** [comment]
@moylop260 thanks, those need to be fixed of course. I meant those: https://travis-ci.org/OCA/server-tools/jobs/161299648#L1730

**Review by hbrunn — APPROVED** [review]
whatever is wrong with travis doesn't seem to be caused by this module

**Review by moylop260 — APPROVED** [review]

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: s/openerp/Odoo

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: s/cpu/CPU

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: What about this?
#. `Settings -> Scheduled Actions` and tune the cron tasks
#. `Settings -> Configuration -> General Settings`
#. Enable Monitor RPC Calls if you want information about RPC calls 
being made (causes some performance overhead).

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: Use capitals for the first letter

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: Please remove this line

File: server_monitoring/README.rst
Comment (elicoidal) [inline server_monitoring/README.rst]: https

File: server_monitoring/__init__.py
Comment (elicoidal) [inline server_monitoring/__init__.py:1]: please use short headers

File: server_monitoring/data/ir_cron.xml
Comment (elicoidal) [inline server_monitoring/data/ir_cron.xml:2]: You can add a copyright notice in the XML files as well

File: server_monitoring/models/server_monitor_database.py
Comment (elicoidal) [inline server_monitoring/models/server_monitor_database.py]: Timestamp

File: server_monitoring/models/server_monitor_database.py
Comment (elicoidal) [inline server_monitoring/models/server_monitor_database.py]: Timestamp (several)

File: server_monitoring/models/server_monitor_database.py
Comment (elicoidal) [inline server_monitoring/models/server_monitor_database.py:70]: Lines read
(to be consistent with the other fields)

File: server_monitoring/models/server_monitor_process.py
Comment (elicoidal) [inline server_monitoring/models/server_monitor_process.py]: s/openerp/Odoo (Several)

File: server_monitoring/models/server_monitor_process.py
Comment (elicoidal) [inline server_monitoring/models/server_monitor_process.py]: Unkown

File: server_monitoring/tests/test_monitor_database.yml
Comment (elicoidal) [inline server_monitoring/tests/test_monitor_database.yml:1]: yml tests being deprecated, should we keep this file?

**Review by elicoidal — CHANGES_REQUESTED** [review]
Nothing blocking except maybe the headers
Thanks for the contribution!

File: server_monitoring/models/server_monitor_process.py
Comment (mtelahun) [inline server_monitoring/models/server_monitor_process.py:78]: I ran into a problem here. You are missing cr and uid arguments to get_param()

**Review by mtelahun — CHANGES_REQUESTED** [review]

File: server_monitoring/models/server_monitor_process.py
Comment (gurneyalex) [reply server_monitoring/models/server_monitor_process.py:78]: fixed, thanks.

**Comment by gurneyalex** [comment]
most of the changes requested by @elicoidal done (I just kept the yaml tests and will rewrite them when porting to odoo 9 or 10, I promise). I think this can be merged :smile_cat:

**Comment by gurneyalex** [comment]
regarding the pylint error on travis, cf. https://github.com/OCA/maintainer-quality-tools/issues/374

regarding the build failure on OCB, I'm baffled. Could there be a weird interaction between this addon and database cleanup? @hbrunn I'd appreciate if you could share some insight...

**Comment by astirpe** [comment]
I've tried this module on a database that is a copy of a db used in production.
All seems working well. I like this module.
:+1:

File: server_monitoring/i18n/server_monitoring.pot
Comment (yajo) [inline server_monitoring/i18n/server_monitoring.pot:1]: Delete pot file.

File: server_monitoring/models/field_types.py
Comment (yajo) [inline server_monitoring/models/field_types.py:20]: Please don't do this. You are polluting the global namespace, no matter if this addon is installed or not, and in all databases in any case.

Just declare the classes and import them from here if you need them, not from `fields`.

And BTW, I'd remove the `osv` version. We want no more old api wherever possible.

File: server_monitoring/models/res_config.py
Comment (yajo) [inline server_monitoring/models/res_config.py:36]: You can do this in a data file, easier.

File: server_monitoring/models/server_monitor_database.py
Comment (yajo) [inline server_monitoring/models/server_monitor_database.py:20]: As said above, this should be in another namespace. OTOH, I think you could simply use `fields.Integer('Row count', readonly=True, type="bigint")` and should simply work. (Not sure eh)

The same applies for all these declarations, so I won't repeat it further down below.

File: server_monitoring/models/server_monitor_database.py
Comment (yajo) [inline server_monitoring/models/server_monitor_database.py:112]: Is Odoo user supposed to have special permissions to do these queries?

If so, please specify that in README and forbid installation with a check in a `pre_install_hook` with a human-readable `ValidationError` raised if there are no perms.

BTW you could return a [`Query`](https://github.com/odoo/odoo/blob/8.0/openerp/osv/query.py#L30) in a separate method for better inheritance. This applies to all literal queries found in addon too.

File: server_monitoring/models/server_monitor_database.py
Comment (yajo) [inline server_monitoring/models/server_monitor_database.py:182]: Do you need this? Can't you use `create` directly?

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:119]: You should use [`_patch_method` (and possibly `_revert_method`)](https://github.com/odoo/odoo/blob/8.0/openerp/models.py#L5211-L5247) instead, as they are the standards for monkey patching.

These methods should be found/returned inside the class to allow inheritance.

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:109]: This is useless.

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:196]: Normally [you should define a lambda calling that method, to allow inheritance, as guidelines explain](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#fields). The same applies in many other places.

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:223]: Put this one at the beginning of class declaration, next to `_name`, `_description` and those kind of vars.

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:104]: You don't need to monkey patch a controller. Just declare it as usually.

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:226]: Are you sure you want this method as public, reachable by XMLRPC?

File: server_monitoring/models/server_monitor_process.py
Comment (yajo) [inline server_monitoring/models/server_monitor_process.py:241]: Are you sure you want this method as public, reachable by XMLRPC?

File: server_monitoring/tests/test_monitor_database.yml
Comment (yajo) [reply server_monitoring/tests/test_monitor_database.yml:1]: Definitely not, although not a blocker for me until they completely disappear.

File: server_monitoring/views/server_monitor_database.xml
Comment (yajo) [inline server_monitoring/views/server_monitor_database.xml:14]: drop version, here and in every form view.

File: server_monitoring/views/server_monitor_database.xml
Comment (yajo) [inline server_monitoring/views/server_monitor_database.xml:77]: String should go into the field itself, right? Here and in other places...

File: server_monitoring/README.rst
Comment (yajo) [inline server_monitoring/README.rst:6]: Fix case please.

File: server_monitoring/README.rst
Comment (yajo) [inline server_monitoring/README.rst:50]: Just curious, are we talking about Prometheus?

**Review by yajo — CHANGES_REQUESTED** [review]
Good idea, indeed, but I wouldn't merge it with current design, honestly. Some changes needed.

File: server_monitoring/models/field_types.py
Comment (gurneyalex) [reply server_monitoring/models/field_types.py:20]: The osv version is required in v8 and probably in v9 too, afair.

File: server_monitoring/README.rst
Comment (gurneyalex) [reply server_monitoring/README.rst:50]: cheap lookalike, maybe yes.

**Comment by gurneyalex** [comment]
can't work on this. Closing the PR.

**Pull request closed without merging**
